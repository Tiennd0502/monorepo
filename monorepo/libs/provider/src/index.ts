export { default as TamaguiProvider } from './TamaguiProvider';
export { default as QueryClientProvider } from './QueryClientProvider';
