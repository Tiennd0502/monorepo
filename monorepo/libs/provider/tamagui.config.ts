import { createTamagui, createTokens } from 'tamagui';
import { config } from '@tamagui/config/v3';
import { themes } from '@tamagui/themes';

import {
  colorsMapping,
  radius,
  size,
  space,
  zIndex,
  defaultFont,
  headingFont,
} from '@monorepo/ui';

const tokens = {
  radius,
  size,
  space,
  zIndex,
};

const tamaguiConfig = createTamagui({
  ...config,
  defaultFont: 'body',
  themes,
  fonts: {
    body: defaultFont,
    heading: headingFont,
  },
  tokens: createTokens({
    ...tokens,
    color: colorsMapping,
  }),
});

type AppConfig = typeof config;

declare module 'tamagui' {
  type TamaguiCustomConfig = AppConfig;
}

export default tamaguiConfig;
