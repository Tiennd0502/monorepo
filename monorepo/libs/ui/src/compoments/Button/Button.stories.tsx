import { Meta, StoryObj } from '@storybook/react';

import Button from '.';

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  args: {
    paddingHorizontal: 20,
  },
};

export default meta;

type Story = StoryObj<typeof Button>;

export const Default: Story = {
  args: {
    children: 'Default',
  },
};

export const OutLined: Story = {
  args: {
    variant: 'outlined',
    children: 'OutLined',
  },
};

export const Disabled: Story = {
  args: {
    disabled: true,
    children: 'Disabled',
  },
};

export const ChromeLess: Story = {
  args: {
    variant: 'chromeless',
    children: 'Chromeless',
  },
};
