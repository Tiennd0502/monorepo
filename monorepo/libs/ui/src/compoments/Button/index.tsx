import { memo } from 'react';
import {
  Button as TButton,
  styled,
  Spinner,
  ButtonProps as TButtonProps,
} from 'tamagui';

const ButtonFrame = styled(TButton, {
  backgroundColor: '$primary',
  justifyContent: 'center',
  alignItems: 'center',
  flexDirection: 'row',
  borderRadius: '$lg',
  borderWidth: 0,
  focusStyle: {
    opacity: 0.7,
  },

  variants: {
    reversal: {
      true: {
        backgroundColor: '$tertiary',
        borderColor: '$tertiary',
      },
    },
    outlined: {
      true: {
        backgroundColor: '$secondary',
        borderColor: '$primary',
        borderWidth: 1,
      },
    },
    disabled: {
      true: {
        backgroundColor: '$backgroundDisabled',
        pointerEvents: 'none',
      },
    },
    size: {
      sm: {
        paddingVertical: '$1.75',
        height: '$9',
        borderRadius: 0,
        borderTopRightRadius: 4,
        borderBottomRightRadius: 4,
      },

      md: {
        paddingVertical: '$3',
        height: '$12.5',
      },

      lg: {
        paddingVertical: '$3',
        height: '$13,5',
      },

      xl: {
        paddingVertical: '$4',
        height: '$15',
      },
    },
  } as const,

  defaultVariants: {
    size: 'md',
  },
});

const ButtonText = styled(TButton.Text, {
  allowFontScaling: false,
  fontFamily: '$heading',
  fontWeight: '600',
  color: '$secondary',

  variants: {
    reversal: {
      true: {
        color: '$secondary',
      },
    },

    outlined: {
      true: {
        color: '$primary',
      },
    },

    disabled: {
      true: {
        opacity: 0.7,
        pointerEvents: 'none',
      },
    },

    size: {
      sm: {
        fontSize: '$textSmall',
        lineHeight: '$2.5',
      },

      md: {
        fontSize: '$textMedium',
        lineHeight: '$4.5',
      },

      lg: {
        fontSize: '$4.5',
        fontWeight: '$4.5',
      },

      xl: {
        fontSize: '$5',
        fontWeight: '$5',
      },
    },
  } as const,
});

interface ButtonProps extends Omit<TButtonProps, 'variant'> {
  isLoading?: boolean;
  variant?: 'outlined' | 'disabled' | 'reversal';
  size?: 'sm' | 'md' | 'lg' | 'xl';
}

const Button = ({
  variant,
  children,
  isLoading = false,
  disabled = false,
  size = 'md',
  ...rest
}: ButtonProps) => {
  const variantObject = variant ? { [variant]: true } : {};

  return (
    <ButtonFrame
      unstyled
      size={size}
      {...variantObject}
      disabled={isLoading || disabled}
      {...rest}
    >
      {isLoading ? (
        <Spinner size="small" color="$tertiary" paddingVertical="$1" />
      ) : (
        <ButtonText
          unstyled
          {...variantObject}
          size={size}
          disabled={isLoading || disabled}
        >
          {children}
        </ButtonText>
      )}
    </ButtonFrame>
  );
};

export default memo(Button);
