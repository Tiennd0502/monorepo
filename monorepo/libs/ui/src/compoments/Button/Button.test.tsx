import { fireEvent, render, waitFor } from '@testing-library/react-native';
import { TamaguiProvider } from '@monorepo/provider';
import Button from '.';

const onPress = jest.fn();

const baseProps = {
  children: 'Click me',
  onPress,
};

const setup = (moreProps = {}) => {
  const props = {
    ...baseProps,
    ...moreProps,
  };
  return render(
    <TamaguiProvider>
      <Button {...props} />
    </TamaguiProvider>
  );
};


describe('Button', () => {

  it('should render button component with default props', () => {
    const container = setup();

    expect(container).toMatchSnapshot();
  });

  it('should render button component with variant outlined', () => {
    const container = setup({variant: 'outlined'});

    expect(container).toMatchSnapshot();
  });

  it('should render button component with props isLoading', () => {
    const container = setup({isLoading: true});

    expect(container).toMatchSnapshot();
  });

  it('should render button component with disabled', () => {
    const container = setup({disabled: true});

    expect(container).toMatchSnapshot();
  });

  it('calls onClick when being clicked', () => {
    const {getByTestId} = setup();

    waitFor(() => {
      const button = getByTestId('button');
  
      fireEvent.press(button);
  
      expect(onPress).toHaveBeenCalled();

    })
    
  });
});
