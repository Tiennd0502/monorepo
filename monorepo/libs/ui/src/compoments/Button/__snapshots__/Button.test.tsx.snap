// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button component with default props 1`] = `
<View
  accessibilityRole="button"
  focusable={true}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#242424",
      "borderBottomLeftRadius": 8,
      "borderBottomRightRadius": 8,
      "borderBottomWidth": 0,
      "borderLeftWidth": 0,
      "borderRightWidth": 0,
      "borderStyle": "solid",
      "borderTopLeftRadius": 8,
      "borderTopRightRadius": 8,
      "borderTopWidth": 0,
      "color": "#fff",
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingTop": 12,
    }
  }
  testID="button"
>
  <Text
    allowFontScaling={false}
    style={
      {
        "color": "#fff",
        "fontFamily": "Gelasio-SemiBold",
        "fontSize": 16,
        "lineHeight": 20,
      }
    }
    suppressHighlighting={true}
  >
    Click me
  </Text>
</View>
`;

exports[`Button should render button component with disabled 1`] = `
<View
  accessibilityRole="button"
  focusVisibleStyle={
    {
      "borderColor": "$background",
    }
  }
  pointerEvents="none"
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#909090",
      "borderBottomLeftRadius": 8,
      "borderBottomRightRadius": 8,
      "borderBottomWidth": 0,
      "borderLeftWidth": 0,
      "borderRightWidth": 0,
      "borderStyle": "solid",
      "borderTopLeftRadius": 8,
      "borderTopRightRadius": 8,
      "borderTopWidth": 0,
      "color": "#fff",
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingTop": 12,
    }
  }
  testID="button"
>
  <Text
    allowFontScaling={false}
    pointerEvents="none"
    style={
      {
        "color": "#fff",
        "fontFamily": "Gelasio-SemiBold",
        "fontSize": 16,
        "lineHeight": 20,
        "opacity": 0.7,
      }
    }
    suppressHighlighting={true}
  >
    Click me
  </Text>
</View>
`;

exports[`Button should render button component with props isLoading 1`] = `
<View
  accessibilityRole="button"
  focusVisibleStyle={
    {
      "borderColor": "$background",
    }
  }
  pointerEvents="none"
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#909090",
      "borderBottomLeftRadius": 8,
      "borderBottomRightRadius": 8,
      "borderBottomWidth": 0,
      "borderLeftWidth": 0,
      "borderRightWidth": 0,
      "borderStyle": "solid",
      "borderTopLeftRadius": 8,
      "borderTopRightRadius": 8,
      "borderTopWidth": 0,
      "color": "#fff",
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingTop": 12,
    }
  }
  testID="button"
>
  <View
    style={
      {
        "flexDirection": "column",
        "paddingBottom": 4,
        "paddingTop": 4,
      }
    }
  >
    <ActivityIndicator
      color=""
      size="small"
    />
  </View>
</View>
`;

exports[`Button should render button component with variant outlined 1`] = `
<View
  accessibilityRole="button"
  focusable={true}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderBottomColor": "#242424",
      "borderBottomLeftRadius": 8,
      "borderBottomRightRadius": 8,
      "borderBottomWidth": 1,
      "borderLeftColor": "#242424",
      "borderLeftWidth": 1,
      "borderRightColor": "#242424",
      "borderRightWidth": 1,
      "borderStyle": "solid",
      "borderTopColor": "#242424",
      "borderTopLeftRadius": 8,
      "borderTopRightRadius": 8,
      "borderTopWidth": 1,
      "color": "#fff",
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingTop": 12,
    }
  }
  testID="button"
>
  <Text
    allowFontScaling={false}
    style={
      {
        "color": "#242424",
        "fontFamily": "Gelasio-SemiBold",
        "fontSize": 16,
        "lineHeight": 20,
      }
    }
    suppressHighlighting={true}
  >
    Click me
  </Text>
</View>
`;
